### ISSUE TITLE
Vendor directory indirectly included in module ZIP due to including `vendor/modules.txt`

### ISSUE THREAD

**Issue by str4d** [golang/go#64440]
### Go version

go version go1.20.5 linux/amd64

### What operating system and processor architecture are you using (`go env`)?

```shell
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/str4d/.cache/go-build"
GOENV="/home/str4d/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/str4d/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/str4d/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.20.5"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1251297887=/tmp/go-build -gno-record-gcc-switches"
```

### What did you do?

I ran the following command to download a module into the module cache:
```
$ go mod download -json "github.com/letsencrypt/boulder@latest"
{
        "Path": "github.com/letsencrypt/boulder",
        "Version": "v0.0.0-20231128193404-e59f59c21bbe",
        "Query": "latest",
        "Info": "/home/str4d/go/pkg/mod/cache/download/github.com/letsencrypt/boulder/@v/v0.0.0-20231128193404-e59f59c21bbe.info",
        "GoMod": "/home/str4d/go/pkg/mod/cache/download/github.com/letsencrypt/boulder/@v/v0.0.0-20231128193404-e59f59c21bbe.mod",
        "Zip": "/home/str4d/go/pkg/mod/cache/download/github.com/letsencrypt/boulder/@v/v0.0.0-20231128193404-e59f59c21bbe.zip",
        "Dir": "/home/str4d/go/pkg/mod/github.com/letsencrypt/boulder@v0.0.0-20231128193404-e59f59c21bbe",
        "Sum": "h1:npLHzIy6Jqp+qenFXPGafBKUhcnwxdN3oo2KSDbjoC4=",
        "GoModSum": "h1:e2DboS7gIS7PlNI2fkfyFB3L1TO6DQ2jlKr6pkgZBXw=",
        "Origin": {
                "VCS": "git",
                "URL": "https://github.com/letsencrypt/boulder",
                "Hash": "e59f59c21bbef9b87388013ab88018c061aabf43"
        }
}
```

I then looked in the module cache for a `vendor` directory.

### What did you expect to see?

```
$ ls /home/str4d/go/pkg/mod/github.com/letsencrypt/boulder@v0.0.0-20231128193404-e59f59c21bbe/vendor
ls: cannot access '/home/str4d/go/pkg/mod/github.com/letsencrypt/boulder@v0.0.0-20231128193404-e59f59c21bbe/vendor': No such file or directory
```

I expected this behaviour due to this rationale section of https://github.com/golang/go/issues/40276:
> ### Why can't vendor directories be used?
> Vendor directories are not included in module zip files. Since they're not
present when a module is downloaded, there's no way to build with them.
>
> We don't plan to include vendor directories in zip files in the future
either. Changing the set of files included in module zip files would break
`go.sum` hashes.

and [this section of the Go Module Reference](https://go.dev/ref/mod#build-commands):
> - By default, if the [go version](https://go.dev/ref/mod#go-mod-file-go) in `go.mod` is 1.14 or higher and a vendor directory is present, the go command acts as if `-mod=vendor` were used. Otherwise, the go command acts as if `-mod=readonly` were used.

and [this section of the Go Module Reference](https://go.dev/ref/mod#vendoring):
> Unlike [vendoring in GOPATH mode](https://go.dev/s/go15vendor), the go command ignores vendor directories in locations other than the main module’s root directory. Additionally, since vendor directories in other modules are not used, the go command does not include vendor directories when building [module zip files](https://go.dev/ref/mod#zip-files) (but see known bugs [#31562](https://go.dev/issue/31562) and [#37397](https://go.dev/issue/37397)).

The [Go Modules Reference](https://go.dev/ref/mod#glos-vendor-directory) defines "vendor directory" as a "directory named `vendor` that contains packages from other modules needed to build packages in the main module".

### What did you see instead?

```
$ ls /home/str4d/go/pkg/mod/github.com/letsencrypt/boulder@v0.0.0-20231128193404-e59f59c21bbe/vendor
modules.txt
```

The vendored modules that are present in the source git repository inside the `vendor` directory are not included in the module ZIP, but the `vendor/modules.txt` files is included, which causes the `vendor` directory to be present. This along with the Go version in `go.mod` being at least 1.14 means that `-mod=vendor` behaviours are triggered by default when running commands like `go list -deps` on the cached module, but unable to function due to the lack of the vendored modules. Working around this requires setting `GOFLAGS="-mod=readonly"` on those commands, to ignore the vendor directory.

If inclusion of `vendor/modules.txt` in module ZIP files is intended behaviour, then the Go Module Reference should be updated to state this, and the definition of "vendor directories" should be updated.

**Comment by mvdan** [comment]
Is this perhaps https://github.com/golang/go/issues/63395?

**Comment by seankhliao** [comment]
Duplicate of #63395

**Comment by str4d** [comment]
Thanks for identifying that! I'd searched the issue tracker and found nothing that seemed relevant before starting to write up the issue, but only figured out it was due to `modules.txt` part way through, which was the keyword I should have searched for.

**Comment by mvdan** [comment]
No worries. I knew the issue existed already, and it still took me a few minutes to find it again :)
